File: Graph/207_Course_Schedule/207_Course_Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        graph = {i: [] for i in range(numCourses)}
        in_degrees = {i: 0 for i in range(numCourses)}
        
        for prerequisite in prerequisites:
            course_to_take = prerequisite[0]
            course_prereq = prerequisite[1]
            graph[course_to_take].append(course_prereq)
            in_degrees[course_prereq] += 1
            
        sources = collections.deque()
        for key in in_degrees:
            if in_degrees[key] == 0:
                sources.append(key)
                
        course_schedules = []
        while sources:
            vertex = sources.popleft()
            course_schedules.append(vertex)
            for neighbor in graph[vertex]:
                in_degrees[neighbor] -= 1
                if in_degrees[neighbor] == 0:
                    sources.append(neighbor)
                    
        return len(course_schedules) == numCourses

File: Graph/207_Course_Schedule/207_Course_Schedule.go
func canFinish(numCourses int, prerequisites [][]int) bool {
    graph := make(map[int][]int)
    in_degree := make(map[int]int)
    
    for i := 0; i < numCourses; i++ {
        in_degree[i] = 0
    }
    
    for _, prerequisite := range prerequisites{
        src, dst := prerequisite[0], prerequisite[1]
        graph[src] = append(graph[src], dst)
        in_degree[dst] += 1
    }

    var queue []int
    for key := range in_degree{
        if in_degree[key] == 0 {
            queue = append(queue, key)
        }
    }

    var course_schedules []int
    for len(queue) > 0 {
        vertex := queue[0]
        queue = queue[1:]
        course_schedules = append(course_schedules, vertex)
        
        for _, child := range graph[vertex] {
            in_degree[child] -= 1
            if in_degree[child] == 0 {
                queue = append(queue, child)
            }
        }
        
    }

    return len(course_schedules) == numCourses
    
}
